### PR TITLE
Fc 649 caching search requests to share between the shared components

### PR DESCRIPTION
**Pull request by ben1stone**

### CODE REVIEW

File: src/Web/Sfa.Das.Sas.Shared.Components.Web/Startup.cs [hunk 1]
```diff
@@ -29,15 +31,22 @@ public void ConfigureServices(IServiceCollection services)
                 options.MinimumSameSitePolicy = SameSiteMode.None;
             });
 
+            services.AddMemoryCache();
```
Comment (desderry) [inline src/Web/Sfa.Das.Sas.Shared.Components.Web/Startup.cs]: this isnt needed with the add fat caching call below

File: src/Web/Sfa.Das.Sas.Shared.Components.Web/Startup.cs [hunk 2]
```diff
@@ -29,15 +31,22 @@ public void ConfigureServices(IServiceCollection services)
...
 
             var fatConfig = new FatSharedComponentsConfiguration();
             Configuration.Bind("fatSharedComponents", fatConfig);
             services.AddSingleton<IFatConfigurationSettings>(fs => fatConfig);
 
+            services.AddSingleton<ICacheStorageService, CacheStorageService>();
+            services.AddScoped<TrainingProviderDetailQueryViewModel>();
```
Comment (desderry) [inline src/Web/Sfa.Das.Sas.Shared.Components.Web/Startup.cs]: these shouldnt be needed here at all first is included in the AddFatSharedComponentsCaching call, im not sure what the second is doing or for!

File: src/Web/Sfa.Das.Sas.Shared.Components/Orchestrators/TrainingProviderOrchestrator.cs [hunk 3]
```diff
@@ -92,28 +96,36 @@ public async Task<TrainingProviderSearchFilterViewModel> GetSearchFilter(Trainin
         public async Task<TrainingProviderDetailsViewModel> GetDetails(TrainingProviderDetailQueryViewModel detailsQueryModel)
         {
 
-            var response = await _mediator.Send(new ApprenticeshipProviderDetailQuery() { UkPrn = Convert.ToInt32(detailsQueryModel.Ukprn), ApprenticeshipId = detailsQueryModel.ApprenticeshipId, ApprenticeshipType = detailsQueryModel.ApprenticeshipType, LocationId = detailsQueryModel.LocationId });
+            var cacheKey = detailsQueryModel.Ukprn + detailsQueryModel.LocationId + detailsQueryModel.ApprenticeshipId;
```
Comment (desderry) [inline src/Web/Sfa.Das.Sas.Shared.Components/Orchestrators/TrainingProviderOrchestrator.cs]: this key needs to be a bit safer, there is a chance multiple objects could have the same key as there is no seperation between the values, i suggest something like 
`var cacheKey = $"providerDetails-{detailsQueryModel.Ukprn}-{detailsQueryModel.LocationId}-{ApprenticeshipId}";`

File: src/Web/Sfa.Das.Sas.Shared.Components/ViewComponents/TrainingProvider/Details/TrainingProviderDetailsViewComponent.cs [hunk 4]
```diff
@@ -1,28 +1,36 @@
...
+using System;
+using SFA.DAS.NLog.Logger;
 
 namespace Sfa.Das.Sas.Shared.Components.ViewComponents.Fat.SearchResults
 {
     public class TrainingProviderDetailsViewComponent : ViewComponent
     {
         private readonly ITrainingProviderOrchestrator _tpOrchestrator;
         private readonly IApprenticeshipOrchestrator _apprenticeshipOrchestrator;
+        private readonly ICacheStorageService _cacheService;
```
Comment (desderry) [inline src/Web/Sfa.Das.Sas.Shared.Components/ViewComponents/TrainingProvider/Details/TrainingProviderDetailsViewComponent.cs]: the cache storage service isnt used here so should be removed

File: src/Web/Sfa.Das.Sas.Shared.Components/ViewModels/TrainingProviderDetailQueryViewModel.cs [hunk 5]
```diff
@@ -11,6 +11,7 @@ public class TrainingProviderDetailQueryViewModel
         public ApprenticeshipType ApprenticeshipType { get; set; }
         public int LocationId { get; set; }
         public string PostCode { get; set; }
+        public TrainingProviderDetailsViewModel TrainingProviderDetails { get; set; }
```
Comment (desderry) [inline src/Web/Sfa.Das.Sas.Shared.Components/ViewModels/TrainingProviderDetailQueryViewModel.cs]: This isnt used anywhere so should be removed

File: src/Web/Sfa.Das.Sas.Web.UnitTests/Infrastructure/Services/CacheStorageServiceTests.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Sfa.Das.Sas.Web.UnitTests.Infrastructure.Services
+{
+    public class CacheStorageServiceTests
```
Comment (desderry) [inline src/Web/Sfa.Das.Sas.Web.UnitTests/Infrastructure/Services/CacheStorageServiceTests.cs]: Need to add some tests around saving to cache

**Comment by desderry** [comment]
i think theres a couple more places which caching needs to happen 1) FATOrchestrator for the searches and 2) in the GetBasketHandler around the framework/standard/provider calls as this will speed up the basket calls

File: src/Web/Sfa.Das.Sas.Infrastructure/Services/CacheStorageService.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Caching.Memory;
+using Newtonsoft.Json;
+
+namespace Sfa.Das.Sas.Infrastructure.Services
+{
+    public class CacheStorageService : ICacheStorageService
```
Comment (desderry) [inline src/Web/Sfa.Das.Sas.Infrastructure/Services/CacheStorageService.cs]: I think this needs to be moved to the ApplicationServices layer too

File: src/Web/Sfa.Das.Sas.Shared.Components/Orchestrators/FatOrchestrator.cs [hunk 8]
```diff
@@ -11,20 +13,38 @@ public class FatOrchestrator : IFatOrchestrator
...
+
+            var model = await _cacheService.RetrieveFromCache<FatSearchResultsViewModel>(cacheKey);
+
+            if (model == null)
+            {
+                var results = _apprenticeshipSearchService.SearchByKeyword(searchQueryModel.Keywords, searchQueryModel.Page, searchQueryModel.ResultsToTake, searchQueryModel.SortOrder, searchQueryModel.SelectedLevels);
+
+                model = _fatSearchResultsViewModelMapper.Map(await results);
 
+                await _cacheService.SaveToCache(cacheKey, model, new TimeSpan(30, 0, 0, 0), new TimeSpan(1, 0, 0, 0));
```
Comment (davidgouge) [inline src/Web/Sfa.Das.Sas.Shared.Components/Orchestrators/FatOrchestrator.cs]: These settings `new TimeSpan(30, 0, 0, 0), new TimeSpan(1, 0, 0, 0)` for the cache expiration appear in 5 places (including here).  Could these be moved out to a constant?

**Pull request merged**